apps/els_lsp/src/els_definition_provider.erl

```diff
@@ -45,3 +52,59 @@ goto_definition(Uri, [POI|Rest]) ->
     _ ->
       goto_definition(Uri, Rest)
   end.
+
+-spec match_incomplete(binary(), pos()) -> [poi()].
+match_incomplete(Text, Pos) ->
```

Comment:
to me these nice heuristics look useful even outside of definition provider, are they?

Would it make sense to gather them into a common place (either into els_parser module or a new els_incomplete_parser module)?

(there is already a section in els_parser that tries to parse incomplete forms. It tries to remove tokens from the end until the thing becomes valid. I guess that is equivalent to a certain kind of `match_before`. Where that function stops parsing, these functions could continue. Im not sure if these should be standalone exported functions that parse a given snippet of a module or should they be part of the generic full module parsing where resulting POIs are stored in the DB)

Reply:
I agree, should probably pull these out to a separate module. Most obviously we could use it for reference provider.